runtime/v2/shim.go

```diff
@@ -176,21 +186,86 @@ func cleanupAfterDeadShim(ctx context.Context, id string, rt *runtime.NSMap[Shim
 
 // ShimInstance represents running shim process managed by ShimManager.
 type ShimInstance interface {
+	io.Closer
+
 	// ID of the shim.
 	ID() string
 	// Namespace of this shim.
 	Namespace() string
 	// Bundle is a file system path to shim's bundle.
 	Bundle() string
-	// Client returns the underlying TTRPC client for this shim.
-	Client() *ttrpc.Client
+	// Client returns the underlying TTRPC or GRPC client object for this shim.
+	// The underlying object can be either *ttrpc.Client or grpc.ClientConnInterface.
+	Client() any
 	// Delete will close the client and remove bundle from disk.
 	Delete(ctx context.Context) error
 }
 
+// makeConnection creates a new TTRPC or GRPC connection object from address.
+// address can be either a socket path for TTRPC or JSON serialized BootstrapParams.
+func makeConnection(ctx context.Context, address string, onClose func()) (_ io.Closer, retErr error) {
+	var (
+		payload = []byte(address)
+		params  client.BootstrapParams
+	)
+
+	if json.Valid(payload) {
+		if err := json.Unmarshal([]byte(address), &params); err != nil {
+			return nil, fmt.Errorf("unable to unmarshal bootstrap params: %w", err)
+		}
+	} else {
+		// Use TTRPC for legacy shims
+		params.Address = address
+		params.Protocol = "ttrpc"
+	}
+
+	switch strings.ToLower(params.Protocol) {
+	case "ttrpc":
+		conn, err := client.Connect(params.Address, client.AnonReconnectDialer)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create TTRPC connection: %w", err)
+		}
+		defer func() {
+			if retErr != nil {
+				conn.Close()
+			}
+		}()
+
+		ttrpcClient := ttrpc.NewClient(conn, ttrpc.WithOnClose(onClose))
+		defer func() {
+			if retErr != nil {
+				ttrpcClient.Close()
+			}
+		}()
+
+		return ttrpcClient, nil
+	case "grpc":
+		var (
+			gopts []grpc.DialOption
+		)
+
+		conn, err := grpc.DialContext(ctx, dialer.DialAddress(params.Address), gopts...)
```

Comment:
Any reason to even pass in gopts if we don't supply any options at the moment?

Reply:
I suspect we'll require some (especially when looking into [containerd client opts](https://github.com/containerd/containerd/blob/20de989afcd2fd4edc20e9b85312e49a8bbe152b/client.go#L122)), I just didn't figure out which ones we need.

Reply:
Possibly a better question, what options should we provide. Do we want to block here? Same send/recv size we use for the daemon clients?

Reply:
Lovely, your message got intertwined with mine before I hit send hahah

Reply:
Right, this is a placeholder for now. I'll follow up with more precise opts configuration.

Reply:
Sounds good